src/dune_rules/watermarks.ml

```diff
@@ -310,20 +308,41 @@ let subst vcs =
           ]
       | Some n -> (n.loc_of_arg, n.arg)
     in
-    match
-      Package.Name.Map.find (Dune_project.packages dune_project.project) name
-    with
-    | Some pkg -> pkg
-    | None ->
-      User_error.raise ~loc
-        [ Pp.textf
-            "Package %s doesn't exist. Please add a (package (name %s) ...) \
-             stanza to your dune-project file."
-            (Package.Name.to_string name)
-            (Package.Name.to_string name)
-        ]
+    let package_named_after_project =
+      let packages = Dune_project.packages dune_project.project in
+      Package.Name.Map.find packages name
+    in
+    let metadata_from_dune_project () =
+      Dune_project.info dune_project.project
+    in
+    let metadata_from_matching_package () =
+      match package_named_after_project with
+      | Some pkg -> Ok pkg.info
+      | None ->
+        Error
+          (User_error.make ~loc
+             [ Pp.textf
+                 "Package %s doesn't exist. Please add a (package (name %s) \
+                  ...) stanza to your dune-project file."
+                 (Package.Name.to_string name)
+                 (Package.Name.to_string name)
+             ])
+    in
+    let version = Dune_project.dune_version dune_project.project in
+    let ok_exn = function
+      | Ok s -> s
+      | Error e -> raise (User_error.E e)
+    in
+    if version >= (3, 0) then
```

Comment:
It seems unusual to refer to dune lang 3.0 here. Maybe we can just add a `(* DUNE3 *)` comment?

Reply:
We do it in other parts of the code as well. If we already have it implemented for the future version, might as well leave it in.